### PR TITLE
Fix remote async with deferred launch policy

### LINKED ISSUE #1523: Remote async with deferred launch policy never executes
Remote `async` does not behave correctly when given a `launch::deferred` policy. The action is not executed by `async` itself nor by waiting on the returned future. The following snippet hangs forever:

    void null_thread() {}
    HPX_PLAIN_ACTION(null_thread, null_action);

    int main() {
        hpx::async<null_action>(hpx::launch::deferred, hpx::find_here()).get();
        return 0;
    }

### PR DESCRIPTION
**Pull request by hkaiser**
- refactored lcos::promise and lcos::packaged_action
- added test

This fixes #1523

### CODE REVIEW

File: hpx/lcos/detail/async_implementations.hpp [hunk 1]
```diff
@@ -41,31 +43,21 @@ namespace hpx { namespace detail
...
+                id1 = naming::id_type(id.get_gid(), naming::id_type::unmanaged);
                 target_is_managed = true;
             }
-            else
-            {
-                p.apply(launch::sync, id, std::forward<Ts>(vs)...);
-            }
+
+            lcos::packaged_action<Action, Result> p;
+            p.apply(std::move(addr), target_is_managed ? id1 : id,
```
Comment (sithhell) [inline hpx/lcos/detail/async_implementations.hpp]: Looks like the launch policy is missing here.

File: hpx/lcos/detail/async_implementations.hpp [hunk 2]
```diff
@@ -85,54 +77,44 @@ namespace hpx { namespace detail
...
                 target_is_managed = true;
             }
-            else
-            {
-                p.apply_cb(launch::sync, id, std::forward<Callback>(cb),
-                    std::forward<Ts>(vs)...);
-            }
+
+            lcos::packaged_action<Action, Result> p;
+            p.apply_cb(std::move(addr), target_is_managed ? id1 : id,
```
Comment (sithhell) [inline hpx/lcos/detail/async_implementations.hpp]: Same here.

File: hpx/lcos/detail/async_implementations.hpp [hunk 3]
```diff
@@ -194,35 +178,39 @@ namespace hpx { namespace detail
...
-                    p.apply(policy, id1, std::forward<Ts>(vs)...);
-                }
-                target_is_managed = true;
-            }
-            else
-            {
-                p.apply(policy, id, std::forward<Ts>(vs)...);
-            }
+            lcos::packaged_action<action_type, result_type> p;
+            p.apply(std::move(addr), target_is_managed ? id1 : id,
```
Comment (sithhell) [inline hpx/lcos/detail/async_implementations.hpp]: The policy is missing.

File: hpx/lcos/detail/future_data.hpp [hunk 4]
```diff
@@ -838,13 +837,11 @@ namespace detail
...
         void set_data(T && result)
         {
-            HPX_ASSERT(started_);
-            this->future_data<Result>::set_value(std::forward<T>(result));
+            this->future_data<Result>::set_data(std::forward<T>(result));
         }
 
         void set_exception(boost::exception_ptr const& e)
         {
-            HPX_ASSERT(started_);
```
Comment (sithhell) [inline hpx/lcos/detail/future_data.hpp]: Why did you remove the assert?

File: hpx/lcos/detail/async_implementations.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/lcos/detail/async_implementations.hpp]: Remote operations don't support launch policies, they never had.

File: hpx/lcos/detail/future_data.hpp (hunk 4 above)
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: The (remote) `lcos::promise` has to support the deferred policy (that's the point of this PR), so it starts the task only when somebody asks for the value. For this reason the task has not started yet.

File: hpx/lcos/detail/async_implementations.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/lcos/detail/async_implementations.hpp]: Am 08.01.2016 11:19 nachm. schrieb "Hartmut Kaiser" <
notifications@github.com>:
>
> In hpx/lcos/detail/async_implementations.hpp:
>
> >                  target_is_managed = true;
> >              }
> > -            else
> > -            {
> > -                p.apply(launch::sync, id, std::forward<Ts>(vs)...);
> > -            }
> > +
> > +            lcos::packaged_action<Action, Result> p;
> > +            p.apply(std::move(addr), target_is_managed ? id1 : id,
>
> Remote operations don't support launch policies, they never had.

If I'm not mistaken, that's for the local case, no? If launch policies are
not supported, why care for deferred?
It sounds a bit contradicting...

**Pull request merged**